api/provider/aws/task_create.go

```diff
@@ -5,29 +5,69 @@ import (
 
 	"github.com/aws/aws-sdk-go/aws"
 	"github.com/aws/aws-sdk-go/service/ecs"
+	"github.com/quintilesims/layer0/common/config"
 	"github.com/quintilesims/layer0/common/models"
 )
 
-// Create runs an ECS Task using the specified Create Task Request. The Create Task
-// Request contains the Task name, the Environment ID, and the Deploy ID.
-// The Deploy ID is used to look up the ECS Task Definition family and version of the
-// Task to run.
+// Create runs an ECS Task using the specified CreateTaskRequest.
+// The CreateTaskRequest contains the ContainerOverrides, the DeployID, the
+// EnvironmentID, the TaskName, and the TaskType.
+//
+// The Deploy ID is used to look up the ECS TaskDefinition Family and Version
+// of the Task to run. The TaskType parameter is one of "stateful" or "stateless"
+// and indicates which ECS LaunchType the user wishes to use ("EC2" or "FARGATE"
+// respectively).
+//
+// Create does not generate any custom errors of its own, but will bubble up errors
+// found in its helper functions as well as errors returned by AWS.
 func (t *TaskProvider) Create(req models.CreateTaskRequest) (string, error) {
 	taskID := entityIDGenerator(req.TaskName)
 	fqEnvironmentID := addLayer0Prefix(t.Config.Instance(), req.EnvironmentID)
+	clusterName := fqEnvironmentID
+	startedBy := t.Config.Instance()
+	taskOverrides := convertContainerOverrides(req.ContainerOverrides)
 
-	deployName, deployVersion, err := lookupDeployNameAndVersion(t.TagStore, req.DeployID)
+	var fargatePlatformVersion string
+	if req.TaskType == models.DeployCompatibilityStateless {
+		fargatePlatformVersion = config.DefaultFargatePlatformVersion
+	}
+
+	taskDefinitionARN, err := lookupTaskDefinitionARNFromDeployID(t.TagStore, req.DeployID)
 	if err != nil {
 		return "", err
 	}
 
-	clusterName := fqEnvironmentID
-	startedBy := t.Config.Instance()
-	taskDefinitionFamily := addLayer0Prefix(t.Config.Instance(), deployName)
-	taskDefinitionVersion := deployVersion
-	taskOverrides := convertContainerOverrides(req.ContainerOverrides)
+	taskDefinition, err := describeTaskDefinition(t.AWS.ECS, taskDefinitionARN)
+	if err != nil {
+		return "", err
+	}
+
+	networkMode := aws.StringValue(taskDefinition.NetworkMode)
+
+	var securityGroupIDs []*string
+	var subnets []string
+	if networkMode == "awsvpc" {
+		environmentSecurityGroupName := getEnvironmentSGName(fqEnvironmentID)
+		environmentSecurityGroup, err := readSG(t.AWS.EC2, environmentSecurityGroupName)
+		if err != nil {
+			return "", err
+		}
+
+		securityGroupIDs = append(securityGroupIDs, environmentSecurityGroup.GroupId)
+
+		subnets = t.Config.PrivateSubnets()
+	}
 
-	task, err := t.runTask(clusterName, startedBy, taskDefinitionFamily, taskDefinitionVersion, taskOverrides)
+	task, err := t.runTask(
+		clusterName,
+		req.TaskType,
```

Comment:
Same things as the service. You're passing in 2 arguments TaskType and fargatePlatformVersion, the latter of which is derived from the former. But fargatePlatformVersion isn't used in the caller's scope.

Reply:
✅